File: go/store.go
package main

import (
	"context"
	"errors"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
	"github.com/jackc/pgx/v5/pgxpool"
)

var (
	ErrNotFound  = errors.New("not found")
	ErrDuplicate = errors.New("duplicate")
)

type User struct {
	ID        int64     `json:"id"`
	Email     string    `json:"email"`
	Name      string    `json:"name"`
	CreatedAt time.Time `json:"created_at"`
}

type OrderItem struct {
	SKU            string `json:"sku"`
	Name           string `json:"name"`
	Qty            int32  `json:"qty"`
	UnitPriceCents int64  `json:"unit_price_cents"`
}

type Order struct {
	ID         int64        `json:"id"`
	UserID     int64        `json:"user_id"`
	Status     string       `json:"status"`
	TotalCents int64        `json:"total_cents"`
	Currency   string       `json:"currency"`
	CreatedAt  time.Time    `json:"created_at"`
	Items      *[]OrderItem `json:"items,omitempty"`
}

const orderCols = "id, user_id, status, total_cents, currency, created_at"

type Store struct {
	db *pgxpool.Pool
}

func scanOrder(row pgx.Row, o *Order) error {
	return row.Scan(&o.ID, &o.UserID, &o.Status, &o.TotalCents, &o.Currency, &o.CreatedAt)
}

func (s *Store) CreateUser(ctx context.Context, email, name string) (*User, error) {
	var u User
	err := s.db.QueryRow(ctx,
		"INSERT INTO users (email, name) VALUES ($1, $2) RETURNING id, email, name, created_at",
		email, name).Scan(&u.ID, &u.Email, &u.Name, &u.CreatedAt)
	var pgErr *pgconn.PgError
	if errors.As(err, &pgErr) && pgErr.Code == "23505" {
		return nil, ErrDuplicate
	}
	return &u, err
}

func (s *Store) GetUser(ctx context.Context, id int64) (*User, error) {
	var u User
	err := s.db.QueryRow(ctx, "SELECT id, email, name, created_at FROM users WHERE id = $1", id).
		Scan(&u.ID, &u.Email, &u.Name, &u.CreatedAt)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, ErrNotFound
	}
	return &u, err
}

func (s *Store) UserExists(ctx context.Context, id int64) (bool, error) {
	var one int
	err := s.db.QueryRow(ctx, "SELECT 1 FROM users WHERE id = $1", id).Scan(&one)
	if errors.Is(err, pgx.ErrNoRows) {
		return false, nil
	}
	return err == nil, err
}

func (s *Store) CreateOrder(ctx context.Context, userID int64, currency string, total int64, items []OrderItem) (*Order, error) {
	tx, err := s.db.Begin(ctx)
	if err != nil {
		return nil, err
	}
	defer tx.Rollback(ctx)

	o := Order{Items: &items}
	err = scanOrder(tx.QueryRow(ctx,
		"INSERT INTO orders (user_id, total_cents, currency) VALUES ($1, $2, $3) RETURNING "+orderCols,
		userID, total, currency), &o)
	if err != nil {
		return nil, err
	}
	skus := make([]string, len(items))
	names := make([]string, len(items))
	qtys := make([]int32, len(items))
	prices := make([]int64, len(items))
	for i, it := range items {
		skus[i], names[i], qtys[i], prices[i] = it.SKU, it.Name, it.Qty, it.UnitPriceCents
	}
	_, err = tx.Exec(ctx, `INSERT INTO order_items (order_id, sku, name, qty, unit_price_cents)
		SELECT $1, * FROM unnest($2::text[], $3::text[], $4::int[], $5::bigint[])`,
		o.ID, skus, names, qtys, prices)
	if err != nil {
		return nil, err
	}
	return &o, tx.Commit(ctx)
}

func (s *Store) GetOrder(ctx context.Context, id int64) (*Order, error) {
	var o Order
	err := scanOrder(s.db.QueryRow(ctx, "SELECT "+orderCols+" FROM orders WHERE id = $1", id), &o)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, ErrNotFound
	}
	if err != nil {
		return nil, err
	}
	rows, err := s.db.Query(ctx,
		"SELECT sku, name, qty, unit_price_cents FROM order_items WHERE order_id = $1 ORDER BY id", id)
	if err != nil {
		return nil, err
	}
	items, err := pgx.CollectRows(rows, pgx.RowToStructByPos[OrderItem])
	if err != nil {
		return nil, err
	}
	o.Items = &items
	return &o, nil
}

func (s *Store) ListOrders(ctx context.Context, userID int64, limit, offset int) ([]Order, error) {
	rows, err := s.db.Query(ctx,
		"SELECT "+orderCols+" FROM orders WHERE user_id = $1 ORDER BY id DESC LIMIT $2 OFFSET $3",
		userID, limit, offset)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	orders := []Order{}
	for rows.Next() {
		var o Order
		if err := scanOrder(rows, &o); err != nil {
			return nil, err
		}
		orders = append(orders, o)
	}
	return orders, rows.Err()
}

File: rust/src/store.rs
use chrono::{DateTime, Utc};
use serde::Serialize;
use sqlx::{FromRow, PgPool};

#[derive(Serialize, FromRow)]
pub struct User {
    pub id: i64,
    pub email: String,
    pub name: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Serialize, FromRow)]
pub struct OrderItem {
    pub sku: String,
    pub name: String,
    pub qty: i32,
    pub unit_price_cents: i64,
}

#[derive(Serialize, FromRow)]
pub struct Order {
    pub id: i64,
    pub user_id: i64,
    pub status: String,
    pub total_cents: i64,
    pub currency: String,
    pub created_at: DateTime<Utc>,
    #[sqlx(skip)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub items: Option<Vec<OrderItem>>,
}

const ORDER_COLS: &str = "id, user_id, status, total_cents, currency, created_at";

pub enum CreateUserError {
    Duplicate,
    Db(sqlx::Error),
}

#[derive(Clone)]
pub struct Store {
    db: PgPool,
}

impl Store {
    pub fn new(db: PgPool) -> Self {
        Self { db }
    }

    pub async fn create_user(&self, email: &str, name: &str) -> Result<User, CreateUserError> {
        sqlx::query_as("INSERT INTO users (email, name) VALUES ($1, $2) RETURNING id, email, name, created_at")
            .bind(email)
            .bind(name)
            .fetch_one(&self.db)
            .await
            .map_err(|e| match &e {
                sqlx::Error::Database(d) if d.is_unique_violation() => CreateUserError::Duplicate,
                _ => CreateUserError::Db(e),
            })
    }

    pub async fn get_user(&self, id: i64) -> sqlx::Result<Option<User>> {
        sqlx::query_as("SELECT id, email, name, created_at FROM users WHERE id = $1")
            .bind(id)
            .fetch_optional(&self.db)
            .await
    }

    pub async fn user_exists(&self, id: i64) -> sqlx::Result<bool> {
        let row: Option<(i32,)> = sqlx::query_as("SELECT 1 FROM users WHERE id = $1")
            .bind(id)
            .fetch_optional(&self.db)
            .await?;
        Ok(row.is_some())
    }

    pub async fn create_order(
        &self,
        user_id: i64,
        currency: &str,
        total: i64,
        items: Vec<OrderItem>,
    ) -> sqlx::Result<Order> {
        let mut tx = self.db.begin().await?;
        let mut order: Order = sqlx::query_as(&format!(
            "INSERT INTO orders (user_id, total_cents, currency) VALUES ($1, $2, $3) RETURNING {ORDER_COLS}"
        ))
        .bind(user_id)
        .bind(total)
        .bind(currency)
        .fetch_one(&mut *tx)
        .await?;

        let skus: Vec<&str> = items.iter().map(|i| i.sku.as_str()).collect();
        let names: Vec<&str> = items.iter().map(|i| i.name.as_str()).collect();
        let qtys: Vec<i32> = items.iter().map(|i| i.qty).collect();
        let prices: Vec<i64> = items.iter().map(|i| i.unit_price_cents).collect();
        sqlx::query(
            "INSERT INTO order_items (order_id, sku, name, qty, unit_price_cents)
             SELECT $1, * FROM unnest($2::text[], $3::text[], $4::int[], $5::bigint[])",
        )
        .bind(order.id)
        .bind(&skus)
        .bind(&names)
        .bind(&qtys)
        .bind(&prices)
        .execute(&mut *tx)
        .await?;
        tx.commit().await?;

        order.items = Some(items);
        Ok(order)
    }

    pub async fn get_order(&self, id: i64) -> sqlx::Result<Option<Order>> {
        let order: Option<Order> = sqlx::query_as(&format!("SELECT {ORDER_COLS} FROM orders WHERE id = $1"))
            .bind(id)
            .fetch_optional(&self.db)
            .await?;
        let Some(mut order) = order else { return Ok(None) };
        let items = sqlx::query_as(
            "SELECT sku, name, qty, unit_price_cents FROM order_items WHERE order_id = $1 ORDER BY id",
        )
        .bind(id)
        .fetch_all(&self.db)
        .await?;
        order.items = Some(items);
        Ok(Some(order))
    }

    pub async fn list_orders(&self, user_id: i64, limit: i64, offset: i64) -> sqlx::Result<Vec<Order>> {
        sqlx::query_as(&format!(
            "SELECT {ORDER_COLS} FROM orders WHERE user_id = $1 ORDER BY id DESC LIMIT $2 OFFSET $3"
        ))
        .bind(user_id)
        .bind(limit)
        .bind(offset)
        .fetch_all(&self.db)
        .await
    }
}
